rlmolecule/alphazero/tfalphazero_problem.py

```diff
@@ -0,0 +1,272 @@
+from abc import abstractmethod
+import itertools
+import logging
+import os
+import time
+from functools import partial
+from pathlib import Path
+from typing import Dict, Optional, Tuple, List
+
+from numpy import array
+
+import sqlalchemy
+import tensorflow as tf
+from tensorflow.python.keras.preprocessing.sequence import pad_sequences
+
+from rlmolecule.alphazero.alphazero_problem import AlphaZeroProblem
+from rlmolecule.alphazero.alphazero_vertex import AlphaZeroVertex
+from rlmolecule.alphazero.keras_utils import KLWithLogits, TimeCsvLogger
+from rlmolecule.tree_search.graph_search_state import GraphSearchState
+
+import tensorflow as tf
+from tensorflow.keras import layers
+
+logger = logging.getLogger(__name__)
+
+
+def make_input_mask(values: tf.Tensor, mask_value: float) -> tf.Tensor:
+    """Masks a given tensor based on values along all but the batch and action axes."""
+    shape = tf.shape(values)
+    values = tf.reshape(values, (shape[0], shape[1], -1))
+    return tf.reduce_all(tf.not_equal(values, mask_value), axis=-1)
+
+def make_action_mask(inputs: List[tf.Tensor], mask_values: list) -> tf.Tensor:
+    """Returns an action mask for a given set of input tensors."""
+    batch_size, max_actions_per_node = tf.shape(inputs[0])[:2]
+    action_mask = tf.constant(True, shape=[batch_size, max_actions_per_node], dtype=bool)
+    for i, inp in enumerate(inputs):
+        new_mask =  make_input_mask(inp, mask_values[i])
+        action_mask = tf.logical_and(action_mask, new_mask)
+    return action_mask
+
+def flatten_batch_and_action_axes(x: tf.Tensor) -> tf.Tensor:
+    """Returns a tensor flattened along the first two axes."""
+    shape = tf.shape(x)
+    return tf.reshape(x, [shape[0] * shape[1], *shape[2:]])
+    
+def build_policy_trainer(model: tf.keras.Model) -> tf.keras.Model:
+    """Returns a wrapper policy model."""
+    value_preds, masked_prior_logits = PolicyWrapper(model)(model.inputs)
+    policy_trainer = tf.keras.Model(model.inputs, [value_preds, masked_prior_logits])
+    return policy_trainer
+
+
+class PolicyWrapper(layers.Layer):
+
+    def __init__(self,
+                 policy_model: tf.keras.Model,
+                 input_masks: dict = {},  # will default to 0 if you don't specify
+                 **kwargs):
+
+        super().__init__(**kwargs)
+        self._policy_model = policy_model
+
+        # Here we create default 0-values for masks, and then update with any 
+        # user supplied masks keyed on the input layer names.
+        self._input_masks = {inp.name: 0. for inp in self._policy_model.inputs}
+        self._input_masks.update(input_masks)
```

Comment:
this looks great -- do we need to worry about dtypes and whether these are floats, numpy, or tf.constants?

Reply:
hmm, the user should definitely be able to supply just a float. but we should definitely cast it correctly.  the default dtype for everything we do is tf.float32, yes?

Reply:
nm, I've been mentally stuck in continuous input land for awhile.  added a small helper function to both set default mask values and cast them according to layer type.